File: android/src/main/java/expo/modules/facecheck/ExpoFaceCheckModule.kt
package expo.modules.facecheck

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.net.Uri
import androidx.exifinterface.media.ExifInterface
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import java.io.File
import java.io.InputStream

private const val DEFAULT_MIN_PIXEL_SIZE: Long = 500_000L
private const val DEFAULT_AREA_THRESHOLD: Double = 0.2

class ExpoFaceCheckModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoFaceCheck")

    AsyncFunction("checkFace") { imageUri: String, options: Map<String, Any?>?, promise: Promise ->
      try {
        val minPixelSize: Long
        val areaThreshold: Double

        val rawMin = options?.get("minPixelSize")
        if (rawMin != null) {
          val value = (rawMin as? Number)?.toDouble()
          if (value == null || value.isNaN() || value.isInfinite() || value < 0) {
            promise.reject(CodedException("ERR_INVALID_OPTIONS", "minPixelSize must be a non-negative finite number", null))
            return@AsyncFunction
          }
          minPixelSize = value.toLong()
        } else {
          minPixelSize = DEFAULT_MIN_PIXEL_SIZE
        }

        val rawArea = options?.get("areaThreshold")
        if (rawArea != null) {
          val value = (rawArea as? Number)?.toDouble()
          if (value == null || value.isNaN() || value.isInfinite() || value < 0.0 || value > 1.0) {
            promise.reject(CodedException("ERR_INVALID_OPTIONS", "areaThreshold must be a number between 0 and 1", null))
            return@AsyncFunction
          }
          areaThreshold = value
        } else {
          areaThreshold = DEFAULT_AREA_THRESHOLD
        }

        val context = appContext.reactContext ?: run {
          promise.reject(CodedException("ERR_NO_CONTEXT", "React context is not available", null))
          return@AsyncFunction
        }

        val uri = Uri.parse(imageUri)

        fun openStream(): InputStream? = when (uri.scheme) {
          "file", null -> File(uri.path ?: imageUri).inputStream()
          "content" -> context.contentResolver.openInputStream(uri)
          else -> File(imageUri).inputStream()
        }

        val decoded = openStream()?.use { BitmapFactory.decodeStream(it) }
        if (decoded == null) {
          promise.reject(CodedException("ERR_LOAD_IMAGE", "Failed to decode image: $imageUri", null))
          return@AsyncFunction
        }

        val orientation = openStream()?.use { ExifInterface(it).getAttributeInt(
          ExifInterface.TAG_ORIENTATION,
          ExifInterface.ORIENTATION_NORMAL
        ) } ?: ExifInterface.ORIENTATION_NORMAL

        val bitmap = applyExifOrientation(decoded, orientation)

        val imageWidth = bitmap.width.toLong()
        val imageHeight = bitmap.height.toLong()

        if (imageWidth * imageHeight < minPixelSize) {
          bitmap.recycle()
          promise.resolve(mapOf(
            "status" to "LOW_QUALITY",
            "faceCount" to 0
          ))
          return@AsyncFunction
        }

        val image = InputImage.fromBitmap(bitmap, 0)

        val detectorOptions = FaceDetectorOptions.Builder()
          .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE)
          .build()

        val detector = FaceDetection.getClient(detectorOptions)

        detector.process(image)
          .addOnSuccessListener { faces ->
            processFaces(faces, areaThreshold, promise)
            bitmap.recycle()
            detector.close()
          }
          .addOnFailureListener { e ->
            promise.reject(CodedException("ERR_FACE_DETECTION", "Face detection failed: ${e.message}", e))
            bitmap.recycle()
            detector.close()
          }
      } catch (e: Exception) {
        promise.reject(CodedException("ERR_FACE_DETECTION", "Face detection error: ${e.message}", e))
      }
    }
  }

  private fun applyExifOrientation(src: Bitmap, orientation: Int): Bitmap {
    val matrix = Matrix()
    when (orientation) {
      ExifInterface.ORIENTATION_ROTATE_90 -> matrix.postRotate(90f)
      ExifInterface.ORIENTATION_ROTATE_180 -> matrix.postRotate(180f)
      ExifInterface.ORIENTATION_ROTATE_270 -> matrix.postRotate(270f)
      ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> matrix.postScale(-1f, 1f)
      ExifInterface.ORIENTATION_FLIP_VERTICAL -> matrix.postScale(1f, -1f)
      ExifInterface.ORIENTATION_TRANSPOSE -> { matrix.postRotate(90f); matrix.postScale(-1f, 1f) }
      ExifInterface.ORIENTATION_TRANSVERSE -> { matrix.postRotate(270f); matrix.postScale(-1f, 1f) }
      else -> return src
    }
    val rotated = Bitmap.createBitmap(src, 0, 0, src.width, src.height, matrix, true)
    if (rotated != src) src.recycle()
    return rotated
  }

  private fun processFaces(faces: List<Face>, areaThreshold: Double, promise: Promise) {
    data class DetectedFace(
      val x: Double,
      val y: Double,
      val width: Double,
      val height: Double,
      val area: Double
    )

    val detected = faces.map { face ->
      val bounds = face.boundingBox
      val w = bounds.width().toDouble()
      val h = bounds.height().toDouble()
      DetectedFace(
        x = bounds.left.toDouble(),
        y = bounds.top.toDouble(),
        width = w,
        height = h,
        area = kotlin.math.abs(w) * kotlin.math.abs(h)
      )
    }.sortedByDescending { it.area }

    if (detected.isEmpty()) {
      promise.resolve(mapOf(
        "status" to "NO_FACE",
        "faceCount" to 0
      ))
      return
    }

    val maxArea = detected[0].area
    val dominant = detected.filter { it.area / maxArea > areaThreshold }
    val dominantCount = dominant.size

    when {
      dominantCount == 0 -> promise.resolve(mapOf(
        "status" to "NO_FACE",
        "faceCount" to 0
      ))
      dominantCount == 1 -> {
        val top = detected[0]
        promise.resolve(mapOf(
          "status" to "READY",
          "faceCount" to 1,
          "dominantFaceBounds" to mapOf(
            "x" to top.x,
            "y" to top.y,
            "width" to top.width,
            "height" to top.height
          )
        ))
      }
      else -> promise.resolve(mapOf(
        "status" to "MULTIPLE_FACES",
        "faceCount" to dominantCount
      ))
    }
  }
}

File: ios/ExpoFaceCheckModule.swift
import ExpoModulesCore
import Vision
import UIKit

private let DEFAULT_MIN_PIXEL_SIZE: CGFloat = 500_000
private let DEFAULT_AREA_THRESHOLD: CGFloat = 0.2

public class ExpoFaceCheckModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoFaceCheck")

    AsyncFunction("checkFace") { (imageUri: String, options: [String: Any]?, promise: Promise) in
      let minPixelSize: CGFloat
      let areaThreshold: CGFloat

      if let raw = options?["minPixelSize"] {
        guard let value = (raw as? NSNumber)?.doubleValue, value.isFinite, value >= 0 else {
          promise.reject("ERR_INVALID_OPTIONS", "minPixelSize must be a non-negative finite number")
          return
        }
        minPixelSize = CGFloat(value)
      } else {
        minPixelSize = DEFAULT_MIN_PIXEL_SIZE
      }

      if let raw = options?["areaThreshold"] {
        guard let value = (raw as? NSNumber)?.doubleValue, value.isFinite, value >= 0, value <= 1 else {
          promise.reject("ERR_INVALID_OPTIONS", "areaThreshold must be a number between 0 and 1")
          return
        }
        areaThreshold = CGFloat(value)
      } else {
        areaThreshold = DEFAULT_AREA_THRESHOLD
      }

      guard let rawImage = self.loadImage(from: imageUri) else {
        promise.reject("ERR_LOAD_IMAGE", "Failed to load image from URI: \(imageUri)")
        return
      }

      guard let image = self.normalizeOrientation(rawImage),
            let cgImage = image.cgImage else {
        promise.reject("ERR_LOAD_IMAGE", "Failed to decode image: \(imageUri)")
        return
      }

      let imageWidth = CGFloat(cgImage.width)
      let imageHeight = CGFloat(cgImage.height)

      if imageWidth * imageHeight < minPixelSize {
        promise.resolve([
          "status": "LOW_QUALITY",
          "faceCount": 0
        ])
        return
      }

      let request = VNDetectFaceRectanglesRequest { request, error in
        if let error = error {
          promise.reject("ERR_FACE_DETECTION", "Face detection failed: \(error.localizedDescription)")
          return
        }

        let results = (request.results as? [VNFaceObservation]) ?? []

        var detectedFaces: [(bounds: CGRect, area: CGFloat)] = []
        for face in results {
          let bbox = face.boundingBox
          let pixelX = bbox.origin.x * imageWidth
          let pixelY = (1.0 - bbox.origin.y - bbox.height) * imageHeight
          let pixelWidth = bbox.width * imageWidth
          let pixelHeight = bbox.height * imageHeight
          let area = abs(pixelWidth) * abs(pixelHeight)
          let pixelBounds = CGRect(x: pixelX, y: pixelY, width: pixelWidth, height: pixelHeight)
          detectedFaces.append((bounds: pixelBounds, area: area))
        }

        if detectedFaces.isEmpty {
          promise.resolve([
            "status": "NO_FACE",
            "faceCount": 0
          ])
          return
        }

        detectedFaces.sort { $0.area > $1.area }
        let maxArea = detectedFaces[0].area
        let dominantCount = detectedFaces.filter { $0.area / maxArea > areaThreshold }.count

        if dominantCount == 0 {
          promise.resolve([
            "status": "NO_FACE",
            "faceCount": 0
          ])
        } else if dominantCount == 1 {
          let dominant = detectedFaces[0]
          promise.resolve([
            "status": "READY",
            "faceCount": 1,
            "dominantFaceBounds": [
              "x": dominant.bounds.origin.x,
              "y": dominant.bounds.origin.y,
              "width": dominant.bounds.width,
              "height": dominant.bounds.height
            ]
          ])
        } else {
          promise.resolve([
            "status": "MULTIPLE_FACES",
            "faceCount": dominantCount
          ])
        }
      }

      let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          try handler.perform([request])
        } catch {
          promise.reject("ERR_FACE_DETECTION", "Failed to perform face detection: \(error.localizedDescription)")
        }
      }
    }
  }

  private func loadImage(from uri: String) -> UIImage? {
    let path: String
    if uri.hasPrefix("file://") {
      guard let url = URL(string: uri) else { return nil }
      path = url.path
    } else if uri.hasPrefix("/") {
      path = uri
    } else {
      guard let url = URL(string: uri), let data = try? Data(contentsOf: url) else { return nil }
      return UIImage(data: data)
    }
    return UIImage(contentsOfFile: path)
  }

  private func normalizeOrientation(_ image: UIImage) -> UIImage? {
    if image.imageOrientation == .up { return image }
    let format = UIGraphicsImageRendererFormat.default()
    format.scale = image.scale
    let renderer = UIGraphicsImageRenderer(size: image.size, format: format)
    return renderer.image { _ in
      image.draw(in: CGRect(origin: .zero, size: image.size))
    }
  }
}
